### PR TITLE
[ML] Refactor ML mappings and templates into JSON resources

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
ML mappings and index templates have so far been created
programmatically. While this had its merits due to static typing,
there is consensus it would be clear to maintain those in json files.
In addition, we are going to adding ILM policies to these indices
and the component for a plugin to register ILM policies is
`IndexTemplateRegistry`. It expects the templates to be in resource
json files.

For the above reasons this commit refactors ML mappings and index
templates into json resource files that are registered via
`MlIndexTemplateRegistry`.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java [hunk 1]
```diff
@@ -57,11 +58,18 @@ public static void loadTemplateIntoMap(String resource, Map<String, IndexTemplat
...
+        return loadTemplate(resource, version, versionProperty, Collections.emptyMap());
+    }
+
+    /**
+     * Loads a built-in template and returns its source after replacing given variables.
+     */
+    public static String loadTemplate(String resource, String version, String versionProperty, Map<String, String> variables) {
         try {
             BytesReference source = load(resource);
+            source = replaceVariables(source, version, versionProperty, variables);
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java]: @gwbrown Here are the changes I made to allow variable replacement on template loading. Could you take a look please when you have some time?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java [hunk 2]
```diff
@@ -887,101 +881,6 @@ protected Clock getClock() {
     public UnaryOperator<Map<String, IndexTemplateMetaData>> getIndexTemplateMetaDataUpgrader() {
         return templates -> {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java]: I have not replaced this with returning an operator that does nothing yet because it is unclear whether `IndexTemplateRegistry` fully addresses upgrading issues. I'm waiting to hear on this, until then the PR is WIP.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProviderTests.java [hunk 3]
```diff
@@ -66,182 +52,20 @@
...
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoMoreInteractions;
 import static org.mockito.Mockito.verifyZeroInteractions;
 import static org.mockito.Mockito.when;
 
 public class JobResultsProviderTests extends ESTestCase {
-    private static final String CLUSTER_NAME = "myCluster";
-
-    @SuppressWarnings("unchecked")
-    public void testCreateJobResultsIndex() {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProviderTests.java]: These tests are unfortunately the result of extreme mocking, something we tried in the past but doesn't seem to bring benefits. I have removed them and replaced them with integration tests in `JobResultsProviderIT`.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProviderTests.java [hunk 4]
```diff
@@ -1031,6 +855,31 @@ public void testDatafeedTimingStats_NotFound() throws IOException {
         verifyNoMoreInteractions(client);
     }
 
+    @SuppressWarnings("unchecked")
+//    public void testTermFieldMapping() throws IOException {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProviderTests.java]: I just noticed this one, I forgot to address it. I'll take care of it before removing the WIP label.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/persistence/MockClientBuilder.java [hunk 5]
```diff
@@ -77,14 +56,11 @@
     private ClusterAdminClient clusterAdminClient;
     private IndicesAdminClient indicesAdminClient;
 
-    private IndicesAliasesRequestBuilder aliasesRequestBuilder;
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/persistence/MockClientBuilder.java]: In hindsight I'm not sure this mock client class was useful. I remove all unused methods.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        ROOT_RESOURCE_PATH + "meta_index_template.json", Version.CURRENT.toString(), VERSION_PATTERN,
+        Collections.singletonMap(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id)));
+
+    private static final IndexTemplateConfig NOTIFICATIONS_TEMPLATE = new IndexTemplateConfig(NotificationsIndex.NOTIFICATIONS_INDEX,
+        ROOT_RESOURCE_PATH + "notifications_index_template.json", Version.CURRENT.toString(), VERSION_PATTERN,
+        Collections.singletonMap(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id)));
+
+    private static final IndexTemplateConfig CONFIG_TEMPLATE = configTemplate();
+
+    private static final IndexTemplateConfig INFERENCE_TEMPLATE = new IndexTemplateConfig(InferenceIndexConstants.LATEST_INDEX_NAME,
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java]: This template's name is formed by the inference prefix plus the latest version. This would mean that once we change version we'll create a new template. I think that's probably not a good idea. I left it as is for now but we can address in this PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateConfig.java [hunk 7]
```diff
@@ -55,7 +82,7 @@ public String getTemplateName() {
      */
     public byte[] loadBytes() {
         String template = TemplateUtils.loadTemplate(fileName, version,
-            Pattern.quote("${" + versionProperty + "}"));
+            Pattern.quote("${" + versionProperty + "}"), variables);
```
Comment (gwbrown) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateConfig.java]: I don't like the difference here between how `versionProperty` is treated (that is, wrapped with `${}` and escaped via `Pattern.quote()`) and how the variables are treated (not wrapped with anything and not escaped) - I think this will lead to confusion, and very possibly accidentally ending up with regexes where we expected to have static strings.

If we don't end up escaping the keys in `variables` either here or in `TemplateUtils#replaceVariable`, it should be made clear that the keys will be compiles as regexes.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java [hunk 8]
```diff
@@ -91,13 +99,22 @@ public static void validate(BytesReference source) {
...
     /**
-     * Filters the source: replaces any template version property with the version number
+     * Replaces all occurences of given variable with the value
      */
-    public static String filter(BytesReference source, String version, String versionProperty) {
-        return Pattern.compile(versionProperty)
-                .matcher(source.utf8ToString())
-                .replaceAll(version);
+    public static String replaceVariable(String input, String variable, String value) {
+        return Pattern.compile(variable)
```
Comment (gwbrown) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java]: Because the only usage of this function is in `TemplateUtils`, maybe we should add the [`LITERAL`](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html#LITERAL) flag here rather than quoting higher in the call stack? I think we're unlikely to need full regex support here.

**Review by AthenaEryma — COMMENTED** [review]
Left a couple comments about escaping in the changes to `TemplateUtils`/`IndexTemplateConfig`, but I think once those are addressed the additional variable support will be valuable.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+public class MlIndexTemplateRegistry extends IndexTemplateRegistry {
+
+    private static final String ROOT_RESOURCE_PATH = "/org/elasticsearch/xpack/core/ml/";
+    private static final String ANOMALY_DETECTION_PATH = ROOT_RESOURCE_PATH + "anomalydetection/";
+    private static final String VERSION_PATTERN = "xpack.ml.version";
+    private static final String VERSION_ID_PATTERN = "xpack.ml.version.id";
+
+    private static final IndexTemplateConfig ANOMALY_DETECTION_RESULTS_TEMPLATE = anomalyDetectionResultsTemplate();
+
+    private static final IndexTemplateConfig ANOMALY_DETECTION_STATE_TEMPLATE = new IndexTemplateConfig(
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java]: We're using the same state index for data frame analytics state too, so maybe drop the `ANOMALY_DETECTION_` bit?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    public MlIndexTemplateRegistry(Settings nodeSettings, ClusterService clusterService, ThreadPool threadPool, Client client,
+                                   NamedXContentRegistry xContentRegistry) {
+        super(nodeSettings, clusterService, threadPool, client, xContentRegistry);
+        this.mlEnabled = MachineLearning.ML_ENABLED.get(nodeSettings);
+    }
+
+    @Override
+    protected List<IndexTemplateConfig> getTemplateConfigs() {
+        if (mlEnabled == false) {
+            return Collections.emptyList();
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java]: This is a difference compared to what we did with the old `getIndexTemplateMetaDataUpgrader()` functionality.  In that we used to return our templates even when ML was disabled on the node.  Returning nothing when ML is disabled isn't necessarily wrong, but it's a difference that needs thinking about.

There may be some subtle effect in Cloud for example.  I think at one time there was a proposal that when a Cloud cluster gets switched to different hardware and a full cluster snapshot from the old hardware is restored into brand new empty cluster running on different hardware that all the X-Pack plugins are disabled during that restore.  I am not sure if this is still being proposed or not.  But it's a scenario where the ML plugin would be disabled in a cluster that contained ML indices.

Let's consult more widely on this offline.

**Review by droberts195 — COMMENTED** [review]
I saw a few nits plus there's a question of what we should do when the plugin is disabled.

Let me know if the test failures are showing up any unforeseen edge cases that need talking through.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateConfig.java (hunk 7 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateConfig.java]: I have pused https://github.com/elastic/elasticsearch/pull/51765/commits/b765370e821caaae07af020a6b54995e673d851c to ensure we wrap and escape everywhere.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java (hunk 8 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java]: Done

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java (hunk 10 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java]: It was wrong to return an empty list. A number of tests was failing because of that. As long as the plugin is loaded, the templates should be there as some nodes may become ML nodes.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java (hunk 9 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java]: I'd like to move all state-related logic in `AnomalyDetectorsIndex` into a `MlStateIndex` class. But I'd rather do that in a follow-up PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java [hunk 11]
```diff
@@ -57,11 +58,18 @@ public static void loadTemplateIntoMap(String resource, Map<String, IndexTemplat
...
+     * Loads a built-in template and returns its source after replacing given variables.
+     */
+    public static String loadTemplate(String resource, String version, String versionProperty, Map<String, String> variables) {
         try {
             BytesReference source = load(resource);
+            source = replaceVariables(source, version, versionProperty, variables);
             validate(source);
-
-            return filter(source, version, versionProperty);
+            return source.utf8ToString();
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java]: there is a lot of conversions and deep copies: utf8->utf16->utf8->utf16. It does not look to hard to implement  `replaceVariables` and `validate` with strings instead of utf8 byte buffers.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java [hunk 12]
```diff
@@ -91,13 +99,22 @@ public static void validate(BytesReference source) {
...
     /**
-     * Filters the source: replaces any template version property with the version number
+     * Replaces all occurences of given variable with the value
      */
-    public static String filter(BytesReference source, String version, String versionProperty) {
-        return Pattern.compile(versionProperty)
-                .matcher(source.utf8ToString())
-                .replaceAll(version);
+    public static String replaceVariable(String input, String variable, String value) {
+        return Pattern.compile("${" + variable + "}", Pattern.LITERAL)
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java]: it looks unsafe to inject a string into a regex pattern like this, it would be good to use `Pattern.quote(variable)` or disallow special characters

**Review by hendrikmuhs — COMMENTED** [review]
Looks good, added 2 comments.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java (hunk 12 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java]: That should be handled by the `LITERAL` flag, no?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java (hunk 11 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java]: Done.

**Comment by gwbrown** [comment]
The changes in https://github.com/elastic/elasticsearch/commit/b765370e821caaae07af020a6b54995e673d851c LGTM. Thank you!

**Comment by dimitris-athanasiou** [comment]
I believe the failing tests in `MlConfigIndexMappingsFullClusterRestartIT` are now because of the lack of template upgrading. This is worked on in https://github.com/elastic/elasticsearch/pull/51456. Once that is merged I should rebase this PR and then I think it'll be ready.

**Comment by dimitris-athanasiou** [comment]
This has now been rebased to pick #51456. Let's see if it passes all the tests.

**Comment by gwbrown** [comment]
Changes in https://github.com/elastic/elasticsearch/pull/51765/commits/0aba76e427518eadb854add3d1d24c5ca64feef1 also LGTM. Thanks again!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java [hunk 13]
```diff
@@ -71,34 +79,43 @@ public static String loadTemplate(String resource, String version, String versio
...
     /**
      * Parses and validates that the source is not empty.
      */
-    public static void validate(BytesReference source) {
+    public static void validate(String source) {
         if (source == null) {
             throw new ElasticsearchParseException("Template must not be null");
         }
+        if (Strings.isEmpty(source)) {
+            throw new ElasticsearchParseException("Template must not be empty");
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/TemplateUtils.java]: nit: I wonder about that: this exception type is used if a REST request is malformed. In this context this makes no sense and we consider the input immutable, right? IMHO this is more of a RTE kind of thing.

Anyway, this is out of scope for this PR, you just add another exception and follow the existing style.

File: x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/core/ml/annotations_index_mappings.json [hunk 14]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/core/ml/annotations_index_mappings.json]: can we put comments in such files? I think it would be good to leave some explanation to the reader of this. I would also add a placeholder for a changelog, in case we use change the schema.

^ applies to all json files introduced

**Review by hendrikmuhs — APPROVED** [review]
LGTM

**Review by droberts195 — APPROVED** [review]
LGTM

Since the CI is green and this change is blocking two other pieces of work from starting I would be happy to merge this as-is and do any additional minor tweaks in followup PRs.

File: x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/core/ml/annotations_index_mappings.json (hunk 14 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/core/ml/annotations_index_mappings.json]: There is no support for comments in JSON files as it stands. We could in theory support them by removing them comment lines when such files are loaded. As for a changelog, I think one of the benefits for those being files, is that we treat them similarly to code files. Thus, git serves as a change log. We can git-blame them and see the commit where each line was introduced/modified. Not sure we need more. But let's see how it goes and we add them if it appears to be needed?
